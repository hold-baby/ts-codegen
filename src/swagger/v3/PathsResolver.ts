import { chain, Dictionary, filter, get, isEmpty, map, pick, reduce } from "lodash";
import { IOperation, IPathItem, IPaths, IReference, IResponse, ISchema, TParameter } from "../v3/OpenAPI";
import { SchemaResolver2 } from "../../SchemaResolver2";

interface IParams {
  pathParams: Array<TParameter | IReference>;
  queryParams: Array<TParameter | IReference>;
  cookieParams: Array<TParameter | IReference>;
}

interface IResolvedPath {
  url: string;
  method: string;
  TResp: any;
  TReq: any;
  operationId?: string;
  params: IParams;
  hasRequestBody: boolean;
}

// TODO: 1. 将 inline 的 requestParams 和 requestBody 抽成单独的 interface，方便外面使用
// TODO: 2. query 不要全部 ...，而是以具体的 {[key]: value} 形式，避免外部应用一些不需要的 query

export class PathsResolver {
  resolver: SchemaResolver2;
  resolvedPaths: IResolvedPath[] = [];
  extraDefinitions: Dictionary<any> = {};

  static of(paths: IPaths, basePath: string = "") {
    return new PathsResolver(paths, basePath);
  }

  constructor(private paths: IPaths, private basePath: string) {
    this.resolver = SchemaResolver2.of((k, v) => {
      if (k) {
        this.extraDefinitions[k] = v;
      }
    });
  }

  scan = () => {
    this.resolvedPaths = reduce(
      this.paths,
      (results: IResolvedPath[], pathItem: IPathItem, pathName: string) => [
        ...results,
        ...this.resolve(pathItem, pathName),
      ],
      [],
    );
    return this;
  };

  resolve(pathItem: IPathItem, pathName: string) {
    const operations = pick(pathItem, ["get", "post", "put", "delete", "patch", "head"]);

    return Object.keys(operations).map((method) => {
      const operation = (operations as Dictionary<IOperation>)[method];
      const params = this.getParams(operation);

      return {
        url: this.getUrl(this.basePath, pathName),
        method,
        operationId: operation.operationId,
        TReq: this.getRequestTypes(params),
        TResp: this.getResponseTypes(operation.responses),
        params: this.getParamsNames(params),
        hasRequestBody: !!operation.requestBody,
      };
    });
  }

  getUrl = (basePath: string, pathName: string) => {
    const path = chain(pathName)
      .split("/")
      .map((p) => (this.isPathParam(p) ? `$${p}` : p))
      .join("/")
      .value();

    return `${basePath}${path === "/" && !!basePath ? "" : path}`;
  };

  toRequestParams = (data: any[] = []) =>
    !isEmpty(data)
      ? `{
    ${data.join(",\n")}
    }`
      : undefined;

  isPathParam = (str: string) => str.startsWith("{");

  // TODO: handle the case when v.parameters = Reference
  getParams = (o: IOperation) => {
    const pickParamsByType = this.pickParams(o.parameters as Array<TParameter | IReference>);

    return {
      pathParams: pickParamsByType("path") as Array<TParameter | IReference>,
      queryParams: pickParamsByType("query") as Array<TParameter | IReference>,
      cookieParams: pickParamsByType("cookie") as Array<TParameter | IReference>,
    };
  };

  getParamsNames = (params: IParams) => {
    const getNames = (list: any[]) => (isEmpty(list) ? [] : map(list, (item) => item.name));
    return {
      pathParams: getNames(params.pathParams),
      queryParams: getNames(params.queryParams),
      cookieParams: [], // TODO: will handle cookieParams later
    };
  };

  getRequestTypes = (params: IParams) => ({
    ...this.getPathParamsTypes(params.pathParams),
    ...this.getQueryParamsTypes(params.queryParams),
  });

  getPathParamsTypes = (pathParams: Array<TParameter | IReference>) => {
    return pathParams.reduce(
      (results, param) => ({
        ...results,
        [`${(param as TParameter).name}${(param as TParameter).required ? "" : "?"}`]: (param as TParameter).type,
      }),
      {},
    );
  };

  getRequestBodyTypes = () => {};

  getQueryParamsTypes = (queryParams: Array<TParameter | IReference>) =>
    queryParams.reduce(
      (o, v) => ({
        ...o,
        [`${(v as TParameter).name}${(v as TParameter).required ? "" : "?"}`]: this.resolver.toType({
          ...(v as ISchema),
          _name: (v as TParameter).name,
          _propKey: (v as TParameter).name,
        }),
      }),
      {},
    );

  // TODO: handle Response or Reference
  // TODO: responses.200.schema.type ==="array"
  // TODO: response.200.schema.type ==="object" (additionalProperties)
  // TODO: response.200.schema.type==="string" | "number" | "integer" | "boolean"
  // TODO: responses.200.headers 存在时
  // TODO: responses.201 同上

  getResponseTypes = (responses: { [responseName: string]: IResponse | IReference }) =>
    this.resolver.toType(get(responses, "200.schema") || get(responses, "201.schema"));

  // TODO: when parameters has enum
  pickParams = (parameters: Array<TParameter | IReference>) => (type: "path" | "query" | "cookie") => {
    return filter(parameters, (param) => {
      // TODO: handle $ref here
      if (param.$ref) {
        return false;
      }
      return (param as TParameter).in === type;
    });
  };
}
