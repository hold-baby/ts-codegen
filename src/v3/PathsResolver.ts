import { chain, compact, Dictionary, filter, get, isEmpty, map, pick, reduce, sortBy } from "lodash";
import { IOperation, IPathItem, IPaths, IReference, IRequestBody, IResponse, TParameter } from "src/v3/OpenAPI";
import { SchemaResolver } from "src/core/SchemaResolver";
import {generateEnums, toTypes} from "src/core/utils";

interface IParams {
  pathParams: Array<TParameter | IReference>;
  queryParams: Array<TParameter | IReference>;
  cookieParams: Array<TParameter | IReference>;
}

interface IResolvedPath {
  url: string;
  method: string;
  TResp: any;
  TReq: any;
  operationId?: string;
  pathParams: string[];
  queryParams: string[];
  cookieParams: string[];
  requestBody: { required: boolean | undefined } | null;
}

// TODO: 1. 将 inline 的 requestParams 和 requestBody 抽成单独的 interface，方便外面使用
// TODO: 2. query 不要全部 ...，而是以具体的 {[key]: value} 形式，避免外部应用一些不需要的 query

export class PathsResolver {
  resolver: SchemaResolver;
  resolvedPaths: IResolvedPath[] = [];
  extraDefinitions: Dictionary<any> = {};

  static of(paths: IPaths, basePath: string = "") {
    return new PathsResolver(paths, basePath);
  }

  constructor(private paths: IPaths, private basePath: string) {
    this.resolver = SchemaResolver.of((k, v) => {
      if (k) {
        this.extraDefinitions[k] = v;
      }
    });
  }

  scan = () => {
    this.resolvedPaths = reduce(
      this.paths,
      (results: IResolvedPath[], pathItem: IPathItem, pathName: string) => [
        ...results,
        ...this.resolve(pathItem, pathName),
      ],
      [],
    );
    return this;
  };

  toRequest = (): string[] => {
    const data = sortBy(this.resolvedPaths, (o) => o.operationId);
    const requests = data.map((v: IResolvedPath) => {
      const TReq = !isEmpty(v.TReq) ? toTypes(v.TReq) : undefined;
      const requestParamList = [
        ...v.pathParams,
        ...v.queryParams,
        ...v.cookieParams,
        v.requestBody ? "requestBody" : undefined,
      ];
      const params = this.toRequestParams(get(v, "queryParams"));

      return `export const ${v.operationId} = createRequestAction<${TReq}, ${v.TResp}>('${v.operationId}', (${
        !isEmpty(requestParamList) ? `${this.toRequestParams(requestParamList)}` : ""
      }) => ({url: \`${v.url}\`, method: "${v.method}", ${v.requestBody ? `data: requestBody,` : ""}${
        params ? `params: ${params},` : ""
      }}));`;
    });

    const enums = Object.keys(this.extraDefinitions).map((k) => generateEnums(this.extraDefinitions, k));
    return [...requests, ...enums];
  };

  resolve(pathItem: IPathItem, pathName: string) {
    const operations = pick(pathItem, ["get", "post", "put", "delete", "patch", "head"]);

    return Object.keys(operations).map((method) => {
      const operation = (operations as Dictionary<IOperation>)[method];
      const params = this.getAllParams(operation);

      return {
        url: this.getUrl(this.basePath, pathName),
        method,
        operationId: operation.operationId,
        TReq: this.getRequestTypes(params, operation.requestBody),
        TResp: this.getResponseTypes(operation.responses),
        ...this.getParamsNames(params),
        requestBody: this.getRequestBody(operation.requestBody),
      };
    });
  }

  getUrl = (basePath: string, pathName: string) => {
    const path = chain(pathName)
      .split("/")
      .map((p) => (this.isPathParam(p) ? `$${p}` : p))
      .join("/")
      .value();

    return `${basePath}${path === "/" && !!basePath ? "" : path}`;
  };

  toRequestParams = (data: any[] = []) =>
    !isEmpty(data)
      ? `{
    ${compact(data).join(",\n")}
    }`
      : undefined;

  isPathParam = (str: string) => str.startsWith("{");

  // TODO: handle the case when v.parameters = Reference
  getAllParams = (o: IOperation) => {
    // TODO: when parameters has enum
    const getParams = (type: "path" | "query" | "cookie") => {
      return filter(o.parameters, (param) => {
        // TODO: handle $ref here
        if (param.$ref) {
          return false;
        }
        return (param as TParameter).in === type;
      });
    };

    return {
      pathParams: getParams("path"),
      queryParams: getParams("query"),
      cookieParams: getParams("cookie"),
    };
  };

  getParamsNames = (params: IParams) => {
    const getNames = (list: any[]) => (isEmpty(list) ? [] : map(list, (item) => item.name));
    return {
      pathParams: getNames(params.pathParams),
      queryParams: getNames(params.queryParams),
      cookieParams: [], // TODO: will handle cookieParams later
    };
  };

  getRequestTypes = (params: IParams, requestBody?: IReference | IRequestBody) => ({
    ...this.getPathParamsTypes(params.pathParams),
    ...this.getQueryParamsTypes(params.queryParams),
    ...this.getRequestBodyTypes(requestBody),
  });

  getPathParamsTypes = (pathParams: Array<TParameter | IReference>) => {
    return pathParams.reduce(
      (results, param) => ({
        ...results,
        [`${(param as TParameter).name}${(param as TParameter).required ? "" : "?"}`]: this.resolver.toType({
          ...(param as TParameter).schema,
          _name: (param as TParameter).name,
          _propKey: (param as TParameter).name,
        }),
      }),
      {},
    );
  };

  getQueryParamsTypes = (queryParams: Array<TParameter | IReference>) =>
    queryParams.reduce(
      (o, v) => ({
        ...o,
        [`${(v as TParameter).name}${(v as TParameter).required ? "" : "?"}`]: this.resolver.toType({
          ...(v as TParameter).schema,
          _name: (v as TParameter).name,
          _propKey: (v as TParameter).name,
        }),
      }),
      {},
    );

  getCookieParamsTypes = (_: Array<TParameter | IReference>) => {};

  getRequestBodyTypes = (requestBody?: IReference | IRequestBody) => {
    if (!requestBody) {
      return "";
    }

    return {
      requestBody: this.resolver.toType((requestBody as IRequestBody).content["application/json"].schema),
    };
  };

  // TODO: handle Response or Reference
  // TODO: responses.200.schema.type ==="array"
  // TODO: response.200.schema.type ==="object" (additionalProperties)
  // TODO: response.200.schema.type==="string" | "number" | "integer" | "boolean"
  // TODO: responses.200.headers 存在时
  // TODO: responses.201 同上

  getResponseTypes = (responses: { [responseName: string]: IResponse | IReference }) => {
    return this.resolver.toType(
      get(responses, [200, "content", "application/json", "schema"]) ||
        get(responses, [201, "content", "application/json", "schema"]),
    );
  };

  getRequestBody = (requestBody?: IReference | IRequestBody) => {
    if (!requestBody) {
      return null;
    }
    // TODO: handle IReference
    if (requestBody.$ref) {
      return null;
    }
    return {
      required: (requestBody as IRequestBody).required,
    };
  };
}
