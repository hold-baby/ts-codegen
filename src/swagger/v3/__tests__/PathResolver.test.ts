import { PathsResolver } from "../PathsResolver";
import swagger from "../../../../examples/swagger.v3.petstore.expanded.json";

describe("PathsResolver", () => {
  it("should get resolved paths by swagger schema", () => {
    expect(PathsResolver.of((swagger as any).paths, "/api/test").scan().resolvedPaths).toEqual(
      expectedPathResolvedData,
    );
  });
});

const expectedPathResolvedData = [
  {
    TReq: {
      "limit?": "number",
      "tags?": "string[]",
    },
    TResp: "IPet[]",
    method: "get",
    operationId: "findPets",
    params: {
      cookieParams: [],
      pathParams: [],
      queryParams: ["tags", "limit"],
    },
    url: "/api/test/pets",
    requestBody: null,
  },
  {
    TReq: {
      requestBody: "INewPet",
    },
    TResp: "IPet",
    method: "post",
    operationId: "addPet",
    params: {
      cookieParams: [],
      pathParams: [],
      queryParams: [],
    },
    url: "/api/test/pets",
    requestBody: {
      required: true,
    },
  },
  {
    TReq: {
      id: "number",
    },
    TResp: "IPet",
    method: "get",
    operationId: "findPetById",
    params: {
      cookieParams: [],
      pathParams: ["id"],
      queryParams: [],
    },
    url: "/api/test/pets/${id}",
    requestBody: null,
  },
  {
    TReq: {
      id: "number",
    },
    TResp: "",
    method: "delete",
    operationId: "deletePet",
    params: {
      cookieParams: [],
      pathParams: ["id"],
      queryParams: [],
    },
    url: "/api/test/pets/${id}",
    requestBody: null,
  },
];
