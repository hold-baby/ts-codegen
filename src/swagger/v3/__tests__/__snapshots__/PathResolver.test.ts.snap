// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PathsResolver should generate correct request from swagger data 1`] = `
"export const addPet = createRequestAction<
  {
    requestBody: INewPet;
  },
  IPet
>(\\"addPet\\", ({ requestBody }) => ({ url: \`/api/test/pets\`, method: \\"post\\", data: requestBody }));

export const deletePet = createRequestAction<{
  id: number;
}>(\\"deletePet\\", ({ id }) => ({ url: \`/api/test/pets/\${id}\`, method: \\"delete\\" }));

export const findPetById = createRequestAction<
  {
    id: number;
  },
  IPet
>(\\"findPetById\\", ({ id }) => ({ url: \`/api/test/pets/\${id}\`, method: \\"get\\" }));

export const findPets = createRequestAction<
  {
    limit?: number;
    tags?: string[];
  },
  IPet[]
>(\\"findPets\\", ({ tags, limit }) => ({
  url: \`/api/test/pets\`,
  method: \\"get\\",
  params: {
    tags,
    limit,
  },
}));
"
`;
